flow/flow.go

```diff
@@ -499,6 +545,167 @@ func (f *Flow) updateMetricsWithNetworkLayer(packet *gopacket.Packet) error {
+		}
+		srcIP = ipv4Packet.SrcIP.String()
+		timeToLive = uint32(ipv4Packet.TTL)
+	case FlowProtocol_IPV6:
+		ipv6Layer := (*packet).Layer(layers.LayerTypeIPv6)
+		ipv6Packet, ok := ipv6Layer.(*layers.IPv6)
+		if !ok {
+			return errors.New("Unable to decode IPv4 Layer")
+		}
+		srcIP = ipv6Packet.SrcIP.String()
+		timeToLive = uint32(ipv6Packet.HopLimit)
+	default:
+		logging.GetLogger().Notice("Capture SYN unknown IP version. ignoring")
+		return nil
+
+	}
+
+	var captureTime int64
+
+	captureTime = f.Last
+	if metadata := (*packet).Metadata(); metadata != nil {
+		captureTime = common.UnixMillis(metadata.CaptureInfo.Timestamp)
+	} else {
+		logging.GetLogger().Debugf("No metadata in packet")
+		return nil
+	}
+
+	switch {
+	case tcpPacket.SYN:
+		var synData []byte
+		if app := (*packet).ApplicationLayer(); app != nil {
+			synData = app.Payload()
+		} else {
+			logging.GetLogger().Notice("Capture SYN failed to parse app data. leaving empty")
+		}
+		if f.Network.A == srcIP {
+			if f.TCPFlowMetric.ABSynStart == 0 {
+				f.TCPFlowMetric.ABSynStart = captureTime
+				f.TCPFlowMetric.ABSynTTL = timeToLive
+				f.TCPFlowMetric.ABSynData = synData
+			} else {
+				logging.GetLogger().Notice("Duplicate SYNCapture AB", f.TCPFlowMetric.ABSynData)
+			}
+		} else {
+			if f.TCPFlowMetric.BASynStart == 0 {
+				f.TCPFlowMetric.BASynStart = captureTime
+				f.TCPFlowMetric.BASynTTL = timeToLive
+				f.TCPFlowMetric.BASynData = synData
+			} else {
+				logging.GetLogger().Notice("Duplicate SYNCapture BA", f.TCPFlowMetric.BASynData)
+			}
+		}
+	case tcpPacket.FIN:
+		if f.Network.A == srcIP {
+			f.TCPFlowMetric.ABFinStart = captureTime
+		} else {
+			f.TCPFlowMetric.BAFinStart = captureTime
+		}
+	case tcpPacket.RST:
+		if f.Network.A == srcIP {
+			f.TCPFlowMetric.ABRstStart = captureTime
+		} else {
+			f.TCPFlowMetric.BARstStart = captureTime
+		}
+	}
+
+	return nil
+}
+
+func (f *Flow) updateMetricsBySeq(packet *gopacket.Packet) error {
+	transportLayer := (*packet).Layer(layers.LayerTypeTCP)
+	transportPacket, transp_ok := transportLayer.(*layers.TCP)
+        if !transp_ok {
+		return nil
+	}
+
+        var expectedSeq uint32
+
+	ipv4Layer := (*packet).Layer(layers.LayerTypeIPv4)
+	ipv4Packet, ok := ipv4Layer.(*layers.IPv4)
+	if ok {
+		if f.Network.A == ipv4Packet.SrcIP.String() {
+			if transportPacket.SYN {
+				f.state.aBExpectedSeq = getExpectedSeq(ipv4Packet, transportPacket)
+				logging.GetLogger().Notice("SYN AB", f.state.aBExpectedSeq)
+				return nil
+			}
+			expectedSeq = f.state.aBExpectedSeq
+		} else {
+			if transportPacket.SYN {
+				f.state.bAExpectedSeq = getExpectedSeq(ipv4Packet, transportPacket)
+				logging.GetLogger().Notice("SYN BA", f.state.bAExpectedSeq)
+				return nil
+			}
+			expectedSeq = f.state.bAExpectedSeq
+		}
+	} else {
+		logging.GetLogger().Error("Error retrieving IP layer")
+	}
+
+	length := uint32Distance(transportPacket.Seq, expectedSeq)
+
+	if transportPacket.Seq < TCPMaxSegmentSize {
```

Comment:
I would prefer to see a logic that track the TCP segment, that will address the TCP partial segment retransmission, etc ...
Add this sliding window logic tracking in flowState
With these information, we could count number of restransmitted and overlaped packets and bytes easily and correctly
In the current version you only retransmission when the restransmit packet if old by TCPMaxSegmentSize, that it different than TCP window size.